### PR TITLE
Use shared statistics merge for union stats

### PR DESCRIPTION
**Pull request by kumarUjjawal**
## Which issue does this PR close?

- Part of #8229

## Rationale for this change

DataFusion already has shared logic for merging `Statistics`, but `UnionExec` and `InterleaveExec` still used their own local merge code.

That left  duplicated path in the codebase and kept the behavior less consistent than the other statistics aggregation paths.

## What changes are included in this PR?

- Reuse `Statistics::try_merge_iter` for `UnionExec` statistics merging
- Reuse the same shared path for `InterleaveExec` statistics merging
- Remove the local union-specific statistics merge helpers
- Add tests for union and interleave statistics merging
- Add a test for interleave partition-level statistics merging

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

### CODE REVIEW

**Comment by kumarUjjawal** [comment]
What's up with the failing avro test? I don't think it is related to this pr.

File: datafusion/physical-plan/src/union.rs [hunk 1]
```diff
@@ -959,294 +903,113 @@ mod tests {
...
+                    .with_null_count(Precision::Exact(1))
+                    .with_byte_size(Precision::Exact(100)),
+            );
 
-        assert_eq!(result, expected);
+        (schema, left, right, expected)
     }
 
     #[test]
-    fn test_union_distinct_count() {
```
Comment (xudong963) [inline datafusion/physical-plan/src/union.rs:1062]: There is a significant concern with the test coverage reduction. 

The old `test_union_distinct_count` contained 15 carefully constructed edge cases for NDV estimation: disjoint ranges, identical ranges, partial overlap, containment, constant values, absent values, and mixed precision types. 

These tests validated the correctness of `estimate_ndv_with_overlap` as used in the union context. The new tests use a single scenario (non-overlapping ranges, single column) and rely on the assumption that `try_merge_iter` is already well-tested elsewhere.

File: testing
Comment (xudong963) [inline testing:1]: is it expected?

**Review by xudong963 — CHANGES_REQUESTED** [review]

File: testing
Comment (kumarUjjawal) [reply testing:1]: This should not have happened, my mistake

File: datafusion/physical-plan/src/union.rs (hunk 1 above)
Comment (kumarUjjawal) [reply datafusion/physical-plan/src/union.rs:1062]: I have plans for the tests in the shared statistics, I am working on adding those.

Comment (xudong963) [reply datafusion/physical-plan/src/union.rs:1062]: nice, I converted the PR to draft now

**Comment by kumarUjjawal** [comment]
@xudong963 Can i get a look on this?

**Review by xudong963 — COMMENTED** [review]
I noticed the NDV fallback change from `sum` to `max` for bound-less inputs is a silent accuracy regression, wdyt

**Comment by kumarUjjawal** [comment]
> I noticed the NDV fallback change from `sum` to `max` for bound-less inputs is a silent accuracy regression, wdyt

Yeah, I changed it.

**Review by xudong963 — APPROVED** [review]
cc @asolimando in case you have a chance to have a look at this.

File: datafusion/common/src/stats.rs [hunk 2]
```diff
@@ -713,7 +713,7 @@ impl Statistics {
                 ) {
                     (Some(&l), Some(&r)) => Precision::Inexact(
                         estimate_ndv_with_overlap(col_stats, item_cs, l, r)
-                            .unwrap_or_else(|| usize::max(l, r)),
+                            .unwrap_or_else(|| l.saturating_add(r)),
```
Comment (asolimando) [inline datafusion/common/src/stats.rs]: The proposed change at this line is a semantic, the proposed fallback is sensible for unions (independent streams, summing NDVs is a good upper bound) but this function is also used to share statistics for Parquet files (see [statistics.rs#L482](https://github.com/apache/datafusion/blob/8a45d0250f3095230902043544a50bdc330131b1/datafusion/datasource/src/statistics.rs#L482) and [statistics.rs#L528](https://github.com/apache/datafusion/blob/8a45d0250f3095230902043544a50bdc330131b1/datafusion/datasource/src/statistics.rs#L528)), for which `max` is a more classic fallback (files from the same table are likely to share common values, so summing NDV would overshoot in general).

One option would be to have a configurable fallback (e.g., an enum `NdvFallback::Max` vs `NdvFallback::Sum`), so the callers can choose based on their own semantics. WDYT?

File: datafusion/physical-plan/src/union.rs (hunk 1 above)
Comment (asolimando) [reply datafusion/physical-plan/src/union.rs:1062]: I think there is still some regression in terms of test coverage:
the old `test_stats_union` covered multi-column merging with mixed types (`Int64`, `Utf8`, `Float32`) and mixed absent/present stats across columns. The new tests use a single `UInt32` column with all stats present.

Could you add a multi-column test case (e.g., 2-3 columns with different types, some with absent stats) to close the gap?

**Review by asolimando — COMMENTED** [review]
Nice cleanup consolidating the union/interleave stats merging onto a single function @kumarUjjawal.

There is still a little gap in test coverage and I think it would be interesting to keep a customizable fallback for NDV merging so the change becomes a pure refactoring with no semantic changes.

Thanks @xudong963 for the ping!

File: datafusion/common/src/stats.rs (hunk 2 above)
Comment (kumarUjjawal) [reply datafusion/common/src/stats.rs]: This sounds like an improvement from the current approach. Thanks @asolimando

**Review by asolimando — APPROVED** [review]
My pending comments were fully addressed, thanks @kumarUjjawal!

EDIT: I noticed there is a [test failure](https://github.com/apache/datafusion/actions/runs/24816808336/job/72632768454?pr=21430), but it seems a flaky test, and unrelated to this PR, so I keep my approval, not sure if guidelines require a green run, in case you can probably re-trigger with an empty commit? (I don't have permission to re-run selectively, and I guess you don't either)

**Comment by kumarUjjawal** [comment]
The CI failure looks lie a flaky EXPLAIN ANALYZE expectation.

**Comment by asolimando** [comment]
> The CI failure looks lie a flaky EXPLAIN ANALYZE expectation.

If you have bandwidth, would you mind filing an issue for this?

**Comment by kumarUjjawal** [comment]
> > The CI failure looks lie a flaky EXPLAIN ANALYZE expectation.
> 
> If you have bandwidth, would you mind filing an issue for this?

I was having issue with this in another pr and I pushed a fix there yesterday, hope that's okay https://github.com/apache/datafusion/pull/21739/changes/94d8f7d6b800cee11d27aaefc68bb3e8e6b024f8

or should i create new issue to address this?

**Comment by asolimando** [comment]
> > > The CI failure looks lie a flaky EXPLAIN ANALYZE expectation.
> > 
> > 
> > If you have bandwidth, would you mind filing an issue for this?
> 
> I was having issue with this in another pr and I pushed a fix there yesterday, hope that's okay [94d8f7d](https://github.com/apache/datafusion/commit/94d8f7d6b800cee11d27aaefc68bb3e8e6b024f8)
> 
> or should i create new issue to address this?

No no, that's fine, I missed that and I wanted to make sure this wouldn't slip through the cracks!

**Comment by kumarUjjawal** [comment]
@xudong963 If this looks good I will merge this now. Can you take a look?

**Comment by xudong963** [comment]
> @xudong963 If this looks good I will merge this now. Can you take a look?

good for me

**Comment by kumarUjjawal** [comment]
Thank you @xudong963 and @asolimando
